### PR TITLE
[Java.time] Calculate week of a year with ISO rules

### LINKED ISSUE #41670: The date_index_name processor calculates wrong week if index_name_format is 'yyyy-w'
**Elasticsearch version** (`bin/elasticsearch --version`):
Version: 7.0.0, Build: default/zip/b7e28a7/2019-04-05T22:55:32.697037Z, JVM: 1.8.0_121

**Plugins installed**: []

**JVM version** (`java -version`):
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):
win7 

**Description of the problem including expected versus actual behavior**:
Indexing a document using the date_index_name processor with `"date_rounding" : "d"` and `"index_name_format": "yyyy-w"` indexes the time `2017-12-04T17:00:58.580` in a index named x-2017-1 while expected index name would be x-2017-49

**Steps to reproduce**:

```
PUT _ingest/pipeline/testingestpipeline
{
  "description": "Pipeline for routing data to specific index by createtime",
  "processors": [
    {
      "date": {
        "field": "createdTime",
        "formats": [
          "date_time","date_time_no_millis", "date_hour_minute_second_millis", "date_hour_minute_second_fraction"
        ]
      },
      "date_index_name": {
        "field": "@timestamp",
        "date_rounding": "d",
        "index_name_prefix": "x-",
        "index_name_format": "yyyy-w"
      }
    }
  ]
}
```

```
DELETE x-*

PUT x-test/_doc/1?pipeline=testingestpipeline
{
  "createdTime": "2017-12-04T17:00:58.580"
}
```
Index name -> **x-2017-1**

If using index_name_format including month and day, the week gets correct

```
PUT _ingest/pipeline/workingingestpipeline
{
  "description": "Pipeline for routing data to specific index by createtime",
  "processors": [
    {
      "date": {
        "field": "createdTime",
        "formats": [
          "date_time","date_time_no_millis", "date_hour_minute_second_millis", "date_hour_minute_second_fraction"
        ]
      },
      "date_index_name": {
        "field": "@timestamp",
        "date_rounding": "d",
        "index_name_prefix": "x-",
        "index_name_format": "yyyy-MM-dd-w"
      }
    }
  ]
}
```

```
DELETE x-*

PUT x-test/_doc/1?pipeline=workingingestpipeline
{
  "createdTime": "2017-12-04T17:00:58.580"
}
```
Index name -> **x-2017-12-04-49**

### PR DESCRIPTION
**Pull request by pgomulka**
IsoLocale - a class introduced to configure start of the week as Monday (with the use of extension) does not allow to configure minimum number of days in a first week. This does not allow correctly calculate week of a week-based-year with ISO rules.
ISO rule is: First day of a week is Monday AND minimum number of days in a first week of a year is 4. 

This change is to refactor this with the use of CalendarDataProvider and requires a jvm property java.locale.providers to be set with SPI,COMPAT. SPI is for this change, COMPAT was already used to be compatible with JDK8 date behaviour. 
Previously not all tests were using COMPAT, so had to be changed now (notice the change in how dates are represented without `.` in text forms)

SQL plugin is at the moment configured to use `SUNDAY` rule - meaning First of the week is Sunday and requires only 1 day in a week to be the first week of the year. This was discussed with SQL team, to keep non-iso functions rely on Sunday,1 rule. cc @astefan  
A link to consider https://docs.microsoft.com/en-us/sql/t-sql/functions/datepart-transact-sql?view=sql-server-ver15#week-and-weekday-datepart-arguments

closes https://github.com/elastic/elasticsearch/issues/41670 a bug raised where week is wrongly calculated
relates https://github.com/elastic/elasticsearch/pull/43652 the previous attempt to just change start of the week

relates #42588 an issue with investigation details
relates #43275 a bug related to week starting on SUNDAY (fixed by #43652 but also by this PR)

### CODE REVIEW

**Comment by pgomulka** [comment]
I think this should be merged to v8 and v7.6. We still support jdk8 in ESv7 and this approach won't work (unless `CalendarDataProvider` is shipped in a separate jar into jre `ext` directory), but it will work for jdk9+

**Comment by pgomulka** [comment]
During the SQL team meeting we decided to keep the current implementation of date functions as they are. However I think they need to be fixed too as I found some problems there. 
I think it is worth visiting https://github.com/pgomulka/elasticsearch/pull/8 (against this branch) before finishing with this current PR.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by pgomulka** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by pgomulka** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by pgomulka** [comment]
@elasticmachine update branch

**Comment by pgomulka** [comment]
@elasticmachine run elasticsearch-ci/oss-distro-docs

File: x-pack/plugin/core/src/test/java/org/elasticsearch/license/LicenseServiceTests.java [hunk 1]
```diff
@@ -17,23 +17,11 @@
...
-        if (expired) {
-            assertThat(message, startsWith("LICENSE [EXPIRED] ON [THU, NOV 15, 2018].\n"));
-        } else {
-            assertThat(message, startsWith("License [will expire] on [Thu, Nov 15, 2018].\n"));
-        }
-    }
-
-    public void testLogExpirationWarningOnJdk8() {
+    public void testLogExpirationWarning() {
         assumeTrue("this is for JDK8 only", JavaVersion.current().equals(JavaVersion.parse("8")));
```
Comment (tvernum) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/license/LicenseServiceTests.java]: @pgomulka I think this part of the change might have been incorrect...
We collapsed the 2 tests into 1, but that one test is set to only run on JDK8. I think this `assumeTrue` should be removed?

Comment (pgomulka) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/license/LicenseServiceTests.java]: @tvernum you are right, the same applies to 7.x branch as the output should be the same.
thanks for taking a look!

**Pull request merged**